shap/maskers/_tabular.py

```diff
@@ -335,3 +335,182 @@ def __init__(self, data, method="linear"):
+            to_be_sampled = np.intersect1d(self.ordering[group_idx], out_of_coalition_indices)
+            if len(to_be_sampled) == 0:
+                continue
+
+            # Condition upon all ancestors in causal ordering
+            to_be_conditioned = np.concatenate(self.ordering[:group_idx])
+            if not self.confounding[group_idx]:
+                # also condition on the in-coalition features from the current causal group
+                in_coalition = np.intersect1d(self.ordering[group_idx], mask)  # does this work?
+                to_be_conditioned = np.union1d(to_be_conditioned, in_coalition).astype(int)
+
+            # Retrieve samples for the features not intervened upon (apply DO-operator)
+            new_samples = self._sample_gaussian(samples, to_be_conditioned, to_be_sampled)
+            samples[:, to_be_sampled] = new_samples
+
+        self._masked_data = samples
+        self._last_mask[:] = mask
+
+        if self.output_dataframe:
+            return pd.DataFrame(self._masked_data, columns=self.feature_names)
+
+        return (self._masked_data,)
+
+    def _sample_gaussian(self, samples, to_be_conditioned, to_be_sampled):
+        # Case 1: No conditioning required (sample marginal distribution)
+        if len(to_be_conditioned) == 0:
+            new_samples = np.random.multivariate_normal(
+                mean=self.mean[to_be_sampled],
+                cov=self.covariance_matrix[np.ix_(to_be_sampled, to_be_sampled)],
+                size=self.n_samples,
+            )
+
+        # Case 2: Conditional sampling (Gaussian conditional distribution)
+        else:
+            # Compute covariance components for conditional Gaussian sampling
+            c = self.covariance_matrix[np.ix_(to_be_sampled, to_be_conditioned)]
+            d = self.covariance_matrix[np.ix_(to_be_conditioned, to_be_conditioned)]
+            cd_inv = c.dot(np.linalg.inv(d))
+            conditional_covariance = self.covariance_matrix[np.ix_(to_be_sampled, to_be_sampled)] - cd_inv.dot(c.T)
+
+            # Ensure covariance matrix symmetry, as required by numpy multivariate_normal
+            if not self._is_symmetric(conditional_covariance):
+                conditional_covariance = self._symmetric_part(conditional_covariance)
+
+            # Compute conditional mean
+            to_sample_means = np.repeat(np.array([self.mean[to_be_sampled]]), self.n_samples, axis=0)
+            to_condition_means = np.repeat(np.array([self.mean[to_be_conditioned]]), self.n_samples, axis=0)
+            conditional_mean = to_sample_means + cd_inv.dot((samples[:, to_be_conditioned] - to_condition_means).T).T
+
+            # Sample
+            new_samples = np.random.multivariate_normal(
+                mean=np.zeros(len(to_be_sampled)), cov=conditional_covariance, size=self.n_samples
+            )
+            new_samples += conditional_mean
+
+        return new_samples
+
+    @staticmethod
+    def _is_positive_definitive(matrix):
```

Comment:
I think it's faster to check whether `matrix - matrix.mean(0)` has `rank == matrix.shape[0]` as per:

https://stats.stackexchange.com/a/53105/206297

But haven't checked.

Reply:
That could be, but since this method is only called in masker initialisation, performance should not really be an issue here. I find the eigenvalue approach clearer, since it is more closely tied to the definition of positive definiteness.